addons/crm/models/crm_lead.py

```diff
@@ -350,6 +350,17 @@ def _compute_name(self):
             if not lead.name and lead.partner_id and lead.partner_id.name:
                 lead.name = _("%s's opportunity") % lead.partner_id.name
 
+    @api.depends('partner_id')
+    def _compute_lang_id(self):
+        lang_codes = self.mapped('partner_id.lang')
+        lang_id_by_code = {
+            lang.code: lang.id
+            for lang in self.env['res.lang'].search([('code', 'in', lang_codes)])
```

Comment:
Note that if you don't have any trithy values, you may skip the search (one query less in 'simple' use cases ^^)